### PR TITLE
Add schema adjustment

### PR DESCRIPTION
**Pull request by k-tomak**
#### Add KustoSinkOption KUSTO_ADJUST_SCHEMA

_Add write option KUSTO_ADJUST_SCHEMA will automatically validate schema inconsistencies between source DataFrame to write and target Kusto table.
KUSTO_ADJUST_SCHEMA behaviour:_
_- If set 'FailIfNotMatch' adjust schemas equality (column name and order) for DataFrame and target Kusto table, fails if schemas don't match._
_- If set 'NoAdjustment' do nothing._
_- If set 'GenerateDynamicCsvMapping', dynamically generates csv mapping based on DataFrame and target Kusto table column names when writing to Kusto. If some Kusto table fields are missing in the DataFrame, they will be ingested as empty. If some DataFrame fields are missing in target table, fails._
_The existing IngestionMappingReference will be reset._
_Default: 'NoAdjustment'_

---

**Features:**
- add KustoSinkOption KUSTO_ADJUST_SCHEMA. 
- add KustoSinkSchemaAdjustmentE2E tests

### CODE REVIEW

**Comment by ohadbitt** [comment]
hey
is this ready for review?

**Comment by k-tomak** [comment]
Yes

чт, 8 июл. 2021 г. в 17:33, ohad bitton ***@***.***>:

**Comment by ohadbitt** [comment]
please merge with latest

**Review by ohadbitt — COMMENTED** [review]
please add validation on 
setMappingOnStagingTableIfNeeded 
that mapping is not configured on the same time that KUSTO_ADJUST_SCHEMA != NoAdjustment

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/SparkIngestionProperties.scala [hunk 1]
```diff
@@ -21,6 +21,19 @@ class SparkIngestionProperties(var flushImmediately: Boolean = false,
...
   }
 
+  def this(ingestionProperties: IngestionProperties){
+    this(
+      ingestionProperties.getFlushImmediately,
+      ingestionProperties.getDropByTags,
+      ingestionProperties.getIngestByTags,
+      ingestionProperties.getAdditionalTags,
+      ingestionProperties.getIngestIfNotExists,
+      null, // no ingestionProperties additionalProperties getter to extract creationTime value
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/SparkIngestionProperties.scala:31]: please refer to my other comment then you wouldnt need this

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoIngestionUtils.scala [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+  }
+
+  private[kusto] def forceAdjustSchema(sourceSchema: StructType, targetSchema: Array[JSONObject]) : Unit ={
+
+    val targetSchemaColumns = targetSchema.map(c=> c.getString("Name")).toSeq
+    val sourceSchemaColumns = sourceSchema.map(c=> c.name)
+
+    if (targetSchemaColumns != sourceSchemaColumns) {
+      throw new RuntimeException(s"Target table schema does not match to DataFrame schema. " +
```
Comment (ohadbitt) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoIngestionUtils.scala:32]: We are moving from using RuntimeException - you can see RetriesExhaustedException for example - lets make a new exceptions folder while we are at it and here call all these exceptions SchemaMatchException or something similar

**Review by ohadbitt — CHANGES_REQUESTED** [review]

File: connector/src/test/scala/com/microsoft/kusto/spark/KustoSinkSchemaAdjustmentE2E.scala [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    spark.sparkContext.stop()
+  }
+
+  override def afterEach(): Unit = {
+    KustoTestUtils.cleanup(kustoConnectionOptions, testTablePrefix)
+  }
+
+
+
+  "Source DataFrame schema adjustment" should "not adjust" in {
```
Comment (ohadbitt) [inline connector/src/test/scala/com/microsoft/kusto/spark/KustoSinkSchemaAdjustmentE2E.scala:44]: tag all tests with 
taggedAs KustoE2E 
our triggered tests currently dont run the E2E

**Comment by k-tomak** [comment]
@ohadbitt I've committed all changes for your review

**Comment by k-tomak** [comment]
About setMappingOnStagingTableIfNeeded. 
IngestionMapping will be updated only if "GenerateDynamicCsvMapping" is set.
When "GenerateDynamicCsvMapping" is set, mappingReferenceName will be reset to Empty, so code block in function setMappingOnStagingTableIfNeeded: 
`if (StringUtils.isNotBlank(mappingReferenceName))`
will be ignored without additional checks.
Or you want to add Exception if user set GenerateDynamicCsvMapping and IngestionMappingReference in SinkOptions simultaneously?

**Comment by ohadbitt** [comment]
> About setMappingOnStagingTableIfNeeded.
> IngestionMapping will be updated only if "GenerateDynamicCsvMapping" is set.
> When "GenerateDynamicCsvMapping" is set, mappingReferenceName will be reset to Empty, so code block in function setMappingOnStagingTableIfNeeded:
> `if (StringUtils.isNotBlank(mappingReferenceName))`
> will be ignored without additional checks.
> Or you want to add Exception if user set GenerateDynamicCsvMapping and IngestionMappingReference in SinkOptions simultaneously?

Yes i want an exception so user is aware that the mapping he provided is ignored

**Comment by k-tomak** [comment]
> Yes i want an exception so user is aware that the mapping he provided is ignored
Done

**Review by ohadbitt — APPROVED** [review]

**Pull request merged**